### PR TITLE
route_lane_change

### PR DESCRIPTION
**Pull request by aswath1**
# PR Details
## Description
As identified in CAR-2785: Create High Level Design for Basic TravelDONE, UI needs a simple topic to display what is the next lane change direction is (left/right) and how far it is until an upcoming lane change is required. This story is to develop this logic, which should leverage existing plugins to identify when the lane change will actually occur and to make this information available to the UI. It is understood that the lane change location cannot be known sooner than it is planned by the relevant plugin. Should be published as soon as it is known and updated as actual point of lane change approaches or shifts.

## Related Issue
https://github.com/usdot-fhwa-stol/carma-platform/issues/1264

## Motivation and Context

lane change direction is (left/right) and how far it is until an upcoming lane change
## Types of changes

- [ ] Defect fix (non-breaking change that fixes an issue)
- [x] New feature (non-breaking change that adds functionality)
- [ ] Breaking change (fix or feature that cause existing functionality to change)

## Checklist:

- [x] I have added any new packages to the sonar-scanner.properties file
- [x] My change requires a change to the documentation.
- [x] I have updated the documentation accordingly.
- [x] I have read the **CONTRIBUTING** document.
[CARMA Contributing Guide](https://github.com/usdot-fhwa-stol/carma-platform/blob/develop/Contributing.md) 
- [x] I have added tests to cover my changes.
- [x] All new and existing tests passed.

### CODE REVIEW

File: .circleci/config.yml [hunk 1]
```diff
@@ -28,7 +28,7 @@ jobs:
     # Pull docker image from docker hub
     # XTERM used for better catkin_make output
     docker:
-      - image: usdotfhwastoldev/autoware.ai:develop
+      - image: usdotfhwastol/autoware.ai:carma-system-3.6.0
```
Comment (dan-du-car) [inline .circleci/config.yml:31]: This change is not needed.

File: Dockerfile [hunk 2]
```diff
@@ -33,7 +33,8 @@
 # Stage 1 - Acquire the CARMA source as well as any extra packages
 # /////////////////////////////////////////////////////////////////////////////
 
-FROM usdotfhwastoldev/autoware.ai:develop AS base-image
+FROM usdotfhwastol/autoware.ai:carma-system-3.6.0 AS base-image
```
Comment (dan-du-car) [inline Dockerfile:36]: This change is not needed

File: carma_wm_ctrl/include/carma_wm_ctrl/WMBroadcaster.h [hunk 3]
```diff
@@ -213,13 +213,6 @@ class WMBroadcaster
    */
   ros::V_string invertParticipants(const ros::V_string& input_participants) const;
```
Comment (dan-du-car) [inline carma_wm_ctrl/include/carma_wm_ctrl/WMBroadcaster.h:215]: need to declare ros::V_string combineParticipantsToVehicle(const ros::V_string& input_participants) const;
undo the change

File: cooperative_lanechange/src/cooperative_lanechange.cpp [hunk 4]
```diff
@@ -794,7 +818,96 @@ namespace cooperative_lanechange
 
     }
 
-     lanelet::BasicLineString2d CooperativeLaneChangePlugin::create_lanechange_path(lanelet::ConstLanelet& start_lanelet,lanelet::ConstLanelet& end_lanelet)
+        for (const auto& p : points)
+        {
```
Comment (dan-du-car) [inline cooperative_lanechange/src/cooperative_lanechange.cpp:822]: This for loop is defined outside the above function, and break the class structure.

File: cooperative_lanechange/src/cooperative_lanechange.cpp [hunk 5]
```diff
@@ -518,42 +518,62 @@ namespace cooperative_lanechange
...
                     ", ending_state_before_buffer_.Y_pos_global" << ending_state_before_buffer_.Y_pos_global);
+            ROS_ERROR_STREAM("ending_state_before_buffer_:"<<ending_state_before_buffer_.X_pos_global << 
+                    ", ending_state_before_buffer_.Y_pos_global" << ending_state_before_buffer_.Y_pos_global);
             
             double route_length = wm_->getRouteEndTrackPos().downtrack;
             
             if(ending_downtrack + ending_buffer_downtrack_ < route_length){
-                ending_pt_index = basic_autonomy::waypoint_generation::get_nearest_index_by_downtrack(route_geometry, wm_, ending_downtrack + ending_buffer_downtrack_);
+                ROS_DEBUG_STREAM("ending_pt_index inside IF calc starting");
+                ending_pt_index = get_ending_point_index(route_geometry, ending_downtrack + ending_buffer_downtrack_);
```
Comment (dan-du-car) [inline cooperative_lanechange/src/cooperative_lanechange.cpp:554]: Not update this

File: cooperative_lanechange/src/cooperative_lanechange.cpp [hunk 6]
```diff
@@ -518,42 +518,62 @@ namespace cooperative_lanechange
...
+           }
+            ROS_ERROR_STREAM("So the latest indexes: nearest_pt_index: " << nearest_pt_index << ", ending_pt_index: " << ending_pt_index);
+            ROS_DEBUG_STREAM("So the latest indexes: nearest_pt_index: " << nearest_pt_index << ", ending_pt_index: " << ending_pt_index);
 
             lanelet::BasicLineString2d future_route_geometry(route_geometry.begin() + nearest_pt_index, route_geometry.begin() + ending_pt_index);
             first = true;
+            ROS_ERROR_STREAM("future geom size:"<<future_route_geometry.size());
             ROS_DEBUG_STREAM("future geom size:"<<future_route_geometry.size());
+            //DEBUG TODO DELETE THIS:
+            if (future_route_geometry.size() == 0 )
```
Comment (dan-du-car) [inline cooperative_lanechange/src/cooperative_lanechange.cpp:568]: Delete it

File: inlanecruising_plugin/include/inlanecruising_plugin/inlanecruising_plugin.h [hunk 7]
```diff
@@ -150,6 +150,32 @@ class InLaneCruisingPlugin
...
+   * \param state The current vehicle state
+   * 
+   * \return index of nearest point in points
+   */ 
+  int get_ending_point_index(const std::vector<lanelet::BasicPoint2d>& points, const cav_msgs::VehicleState& state) const;
+
+  /**
+   * \brief Helper method to split a list of PointSpeedPair into separate point and speed lists 
+   */ 
+  void split_point_speed_pairs(const std::vector<PointSpeedPair>& points, std::vector<lanelet::BasicPoint2d>* basic_points,
```
Comment (dan-du-car) [inline inlanecruising_plugin/include/inlanecruising_plugin/inlanecruising_plugin.h:176]: Please refer to the latest release branch as well

File: route_following_plugin/include/route_following_plugin.h [hunk 8]
```diff
@@ -24,6 +24,7 @@
 #include <carma_wm/WMListener.h>
 #include <carma_wm/WorldModel.h>
 #include <cav_srvs/PlanManeuvers.h>
+#include <cav_msgs/UpcomingLaneChangeStatus.h>
```
Comment (dan-du-car) [inline route_following_plugin/include/route_following_plugin.h:27]: Is this downtrack_until_lanechange unit of measure mile or meters?

File: carma/launch/guidance.launch [hunk 9]
```diff
@@ -109,8 +109,10 @@
   <include file="$(find pure_pursuit_jerk_wrapper)/launch/pure_pursuit_jerk_wrapper.launch" />
 
     <!-- Route Following Plugin -->
-  <include file="$(find route_following_plugin)/launch/route_following_plugin.launch" />
-  
+  <group>
+    <include file="$(find route_following_plugin)/launch/route_following_plugin.launch" />
+    <remap from="upcoming_lane_change_status" to="$(optenv CARMA_GUIDE_NS)/upcoming_lane_change_status"/>
```
Comment (msmcconnell) [inline carma/launch/guidance.launch:114]: As mentioned in the other PR having this remapping below the include line does nothing. If it is needed, move it above. Otherwise delete it

File: carma_wm_ctrl/include/carma_wm_ctrl/WMBroadcaster.h [hunk 10]
```diff
@@ -268,6 +261,9 @@ class WMBroadcaster
    * It should not be incremented for updates that do not require a full map publication.
    */
   size_t current_map_version_ = 0; 
+  ros::V_string combineParticipantsToVehicle(const ros::V_string& input_participants) const;
```
Comment (msmcconnell) [inline carma_wm_ctrl/include/carma_wm_ctrl/WMBroadcaster.h:264]: This function was moved but its method comments were deleted. Please add the comments back. It also looks like it was moved from public to private. Please confirm this does not cause build issues

File: carma_wm_ctrl/src/WMBroadcaster.cpp [hunk 11]
```diff
@@ -573,7 +573,7 @@ lanelet::ConstLaneletOrAreas WMBroadcaster::getAffectedLaneletOrAreas(const cav_
     prev_pt.x += pt.x;
     prev_pt.y += pt.y;
 
-    ROS_DEBUG_STREAM("After conversion in Map frame: Point X "<< gf_pts.back().x() <<" After conversion: Point Y "<< gf_pts.back().y());
+    ROS_DEBUG_STREAM("After conversion: Point X "<< gf_pts.back().x() <<" After conversion: Point Y "<< gf_pts.back().y());
```
Comment (msmcconnell) [inline carma_wm_ctrl/src/WMBroadcaster.cpp:576]: Why were the more descriptive "in TCM frame" and "in Map frame" parts of these comments removed? Seems arbitrary and provides less information in the print line.

File: cooperative_lanechange/src/cooperative_lanechange.cpp [hunk 12]
```diff
@@ -61,8 +61,6 @@ namespace cooperative_lanechange
         twist_sub_ = nh_->subscribe("current_velocity", 1, &CooperativeLaneChangePlugin::twist_cd, this);
         incoming_mobility_response_ = nh_->subscribe("incoming_mobility_response", 1 , &CooperativeLaneChangePlugin::mobilityresponse_cb, this);
         
-        georeference_sub_ = nh_->subscribe("georeference", 1, &CooperativeLaneChangePlugin::georeference_callback, this);
```
Comment (msmcconnell) [inline cooperative_lanechange/src/cooperative_lanechange.cpp:64]: The entire change set in this file is seems to be reverting the GPS fix which I completed last week. It seems to suggest an incorrect merge. If you don't know how to properly merge something you need to ask for assistance. Accidently reverting fixes is extremely problematic.

File: inlanecruising_plugin/config/parameters.yaml [hunk 13]
```diff
@@ -6,7 +6,7 @@ turn_downsample_ratio: 20 # Amount to downsample input lanelet centerline data o
 minimum_speed: 2.2352 # Minimum allowable speed in m/s
 speed_moving_average_window_size: 5 # Size of the window used in the moving average filter to smooth the output speeds
 curvature_moving_average_window_size: 9 # Size of the window used in the moving average filter to smooth the computed curvature
-back_distance: 20.0 # Number of meters behind the first maneuver that need to be included in points for curvature calculation
+back_distance: 30.0 # Number of meters behind the first maneuver that need to be included in points for curvature calculation
```
Comment (msmcconnell) [inline inlanecruising_plugin/config/parameters.yaml:9]: Why was this changed? This seems totally unrelated to the feature added by this PR

File: inlanecruising_plugin/include/inlanecruising_plugin/inlanecruising_config.h [hunk 14]
```diff
@@ -39,7 +39,7 @@ struct InLaneCruisingPluginConfig
   double back_distance = 20;               // Number of meters behind the first maneuver that need to be included in points for curvature calculation
   bool enable_object_avoidance = true;    // Activate object avoidance logic
   bool publish_debug = false; // True if debug publishing will be enabled
-  double buffer_ending_downtrack = 20.0;
+  double buffer_ending_downtrack = 10.0;
```
Comment (msmcconnell) [inline inlanecruising_plugin/include/inlanecruising_plugin/inlanecruising_config.h:42]: Same as above. This is an unrelated change for this feature.

File: inlanecruising_plugin/include/inlanecruising_plugin/inlanecruising_plugin.h [hunk 15]
```diff
@@ -150,6 +150,32 @@ class InLaneCruisingPlugin
                                                   double max_starting_downtrack,
                                                   const carma_wm::WorldModelConstPtr& wm);
 
+  /**
```
Comment (msmcconnell) [inline inlanecruising_plugin/include/inlanecruising_plugin/inlanecruising_plugin.h:153]: All the changes in this folder seem unrelated to PR feature set. This seems like an invalid merge

File: inlanecruising_plugin/test/test_inlanecruising_plugin.cpp [hunk 16]
```diff
@@ -204,7 +204,7 @@ TEST(InLaneCruisingPluginTest, get_nearest_index_by_downtrack_test)
   ASSERT_EQ(3, basic_autonomy::waypoint_generation::get_nearest_index_by_downtrack(points, wm, state));
   ASSERT_EQ(3, basic_autonomy::waypoint_generation::get_nearest_index_by_downtrack(basic_points, wm, state));
 }
-
+*/
```
Comment (msmcconnell) [inline inlanecruising_plugin/test/test_inlanecruising_plugin.cpp:207]: Why was this unit test commented out? Why was it even edited?

File: pure_pursuit_wrapper/include/pure_pursuit_wrapper/pure_pursuit_wrapper.hpp [hunk 17]
```diff
@@ -52,13 +52,6 @@ class PurePursuitWrapper {
          */ 
         std::vector<double> apply_response_lag(const std::vector<double>& speeds, const std::vector<double> downtracks, double response_lag) const;
 
-        /**
-         * \brief Drops any points that sequentially have same target_time and return new trajectory_points in order to avoid divide by zero situation
-         * \param traj_points Velocity profile to shift. The first point should be the current vehicle speed
-         * 
-         * NOTE: This function assumes the target_time will not go backwards. In other words, it only removes "sequential" points that have same target_time
-         * \return A new trajectory without any repeated time_stamps
-         */
```
Comment (msmcconnell) [inline pure_pursuit_wrapper/include/pure_pursuit_wrapper/pure_pursuit_wrapper.hpp:61]: Helpful function comments should not be deleted

**Review by msmcconnell — CHANGES_REQUESTED** [review]
Before I continue, its very clear that this PR contains a faulty merge change set with develop, as numerous files are edited which should have no impact on the feature set represented by this PR. 
This must be resolved before review/merge can continue.

File: route_following_plugin/src/route_following_plugin.cpp [hunk 18]
```diff
@@ -192,6 +214,62 @@ namespace route_following_plugin
...
+            ", downtrack until that lanechange: " << upcoming_lane_change_status_msg_map_.front().first);
+            upcoming_lane_change_status_msg_map_.front().second.downtrack_until_lanechange=upcoming_lane_change_status_msg_map_.front().first-current_progress;
+            ROS_DEBUG_STREAM("upcoming_lane_change_status_msg_map_.front().second.downtrack_until_lanechange: " <<static_cast<double>(upcoming_lane_change_status_msg_map_.front().second.downtrack_until_lanechange));
+            upcoming_lane_change_status_pub_.publish(upcoming_lane_change_status_msg_map_.front().second); 
+        }
+        //DEBUG PURPOSE ONLY, DELETE AFTER TESTING
+        if (upcoming_lane_change_status_msg_map_.empty())
+        {
+            ROS_DEBUG_STREAM("lanechange status map is empty");
+        }
```
Comment (msmcconnell) [inline route_following_plugin/src/route_following_plugin.cpp:270]: Based on the above comment this if statement can be deleted

File: route_following_plugin/src/route_following_plugin.cpp [hunk 19]
```diff
@@ -104,30 +103,47 @@ namespace route_following_plugin
...
+
+                //Determine the Lane Change Status
+                ROS_DEBUG_STREAM("Recording lanechange start_dist <<" << start_dist  << ", from llt id:" << route_shortest_path[shortest_path_index].id() << " to llt id: " << 
+                    route_shortest_path[shortest_path_index+ 1].id());
+                upcoming_lane_change_status_msg_map_.push({start_dist, ComposeLaneChangeStatus(route_shortest_path[shortest_path_index],route_shortest_path[shortest_path_index + 1])});
             }
             else
             {
+                ROS_DEBUG_STREAM("Lanechange NOT Needed ");
+                ROS_ERROR_STREAM("Lanechange NOT Needed :");
```
Comment (msmcconnell) [inline route_following_plugin/src/route_following_plugin.cpp:146]: All these duplicate ERROR logs should be removed as they will clog up our log files and are not in fact errors. The debug logs are enough as they can be trivially enabled in carma-config when needed.

File: route_following_plugin/src/route_following_plugin.cpp [hunk 20]
```diff
@@ -161,17 +177,22 @@ namespace route_following_plugin
...
+
                 ++i;
                 continue;
             }
             if(planned_time == 0.0){
                 //update start distance of first maneuver
                 setManeuverStartDist(latest_maneuver_plan_[i], current_downtrack);
             }
-            planned_time += getManeuverDuration(latest_maneuver_plan_[i], epsilon_).toSec();
+            planned_time += getManeuverDuration(latest_maneuver_plan_[i], 0.0001).toSec();
```
Comment (msmcconnell) [inline route_following_plugin/src/route_following_plugin.cpp:195]: Seems like the epsilon_ constant got replaced with its actual value? Probably best to change it back since the epsilon_ variable at least had a comment explaining its use.

File: route_following_plugin/include/route_following_plugin.h [hunk 21]
```diff
@@ -64,6 +65,18 @@ namespace route_following_plugin
          */
         void initialize();
 
+        // wm listener pointer and pointer to the actual wm object
+        std::shared_ptr<carma_wm::WMListener> wml_;
+        carma_wm::WorldModelConstPtr wm_;
+
+        /**
+         * \brief Compose UpcomingLaneChangeStatus msg from given starting and ending lanelets
```
Comment (msmcconnell) [inline route_following_plugin/include/route_following_plugin.h:73]: This method comment should specify that this method will only function correctly if the two provided lanelets are actually forming a lane change. This is an important assumption to document

File: route_following_plugin/config/parameters.yaml [hunk 22]
```diff
@@ -3,7 +3,7 @@ minimal_plan_duration: 15
 
 #Double : The jerk used to bring vehicle to a halt at end of route
 #Unit : m/s3
-route_end_jerk : 3.0
+route_end_jerk : 100.0
```
Comment (msmcconnell) [inline route_following_plugin/config/parameters.yaml:6]: Is this change intentional or from an invalid merge?

File: route_following_plugin/src/route_following_plugin.cpp [hunk 23]
```diff
@@ -192,6 +214,62 @@ namespace route_following_plugin
...
+    }
+
+    void RouteFollowingPlugin::pose_cb(const geometry_msgs::PoseStampedConstPtr& msg)
+    {
+
+        ROS_DEBUG_STREAM("Entering pose_cb");
+        pose_msg_ = geometry_msgs::PoseStamped(*msg.get());
+
+        if (!wm_->getRoute())
+        return;
```
Comment (msmcconnell) [inline route_following_plugin/src/route_following_plugin.cpp:244]: indent this line as it is very easy to misread

File: unobstructed_lanechange/include/unobstructed_lanechange.h [hunk 24]
```diff
@@ -80,6 +80,8 @@ namespace unobstructed_lanechange
                                                 double max_starting_downtrack,
                                                 const carma_wm::WorldModelConstPtr& wm,const cav_msgs::VehicleState& state);
             int getNearestRouteIndex(lanelet::BasicLineString2d& points, const cav_msgs::VehicleState& state);
+
+            int getNearestPointIndex(const std::vector<lanelet::BasicPoint2d>& points, const cav_msgs::VehicleState& state) const;
```
Comment (msmcconnell) [inline unobstructed_lanechange/include/unobstructed_lanechange.h:84]: @lewisid Are these changes to unobstructed_lanechange intentional or from a bad merge? If the intentional, I'll need to get Anish to review.

**Review by msmcconnell — CHANGES_REQUESTED** [review]
Looks like most of the merge issues are resolved, though there is one or two files of concern still

File: route_following_plugin/src/route_following_plugin.cpp [hunk 25]
```diff
@@ -104,30 +101,40 @@ namespace route_following_plugin
...
             start_speed = (maneuvers.empty()) ? 0.0 : GET_MANEUVER_PROPERTY(maneuvers.back(), end_speed);
+            ROS_DEBUG_STREAM("start_dist:" << start_dist << ", start_speed:" << start_speed);
 
+                
             end_dist = wm_->routeTrackPos(route_shortest_path[shortest_path_index].centerline2d().back()).downtrack;
+            ROS_DEBUG_STREAM("end_dist:" << end_dist);
+
             end_dist = std::min(end_dist, route_length);
-            
-            if (std::fabs(start_dist - end_dist) < 0.1) //TODO: edge case that was not recreatable. Sometimes start and end dist was same which crashes inlanecruising
```
Comment (msmcconnell) [inline route_following_plugin/src/route_following_plugin.cpp:117]: Please add this if block back. Unless you know that whatever it prevented was resolved

File: route_following_plugin/src/route_following_plugin.cpp [hunk 26]
```diff
@@ -149,7 +156,6 @@ namespace route_following_plugin
     {
         if (latest_maneuver_plan_.empty())
         {
-            ROS_ERROR_STREAM("A maneuver plan has not been generated");
```
Comment (msmcconnell) [inline route_following_plugin/src/route_following_plugin.cpp:152]: Why did you delete this line? It was not a duplicate. Please think through each change with a bit more diligence.

**Review by msmcconnell — CHANGES_REQUESTED** [review]

File: unobstructed_lanechange/config/parameters.yaml [hunk 27]
```diff
@@ -65,4 +65,8 @@ enable_object_avoidance_lc : false
 
 # Double: The minimum timestep used for planning trajectory
 #Units: seconds
-min_timestep : 0.1
\ No newline at end of file
+min_timestep : 0.1
+
+# Double : The additional downtrack beyond requested end dist used to fit points along spline. Extra points removed before returning trajectory. Feature not used if end dist is route end
+#Units: meters
+ending_buffer_downtrack : 5.0
```
Comment (msmcconnell) [inline unobstructed_lanechange/config/parameters.yaml:72]: Still have unobstructed_lane change package changes in this and the launch file.
The launch file change actually looks correct, but I don't know if its functional. Please check with @adev4a

**Review by msmcconnell — APPROVED** [review]
